src/main/java/com/sarality/sync/APISyncExecutor.java

```diff
@@ -41,56 +48,78 @@ public void execute() {
     execute(null, null);
   }
 
-  public void execute(TaskProgressListener<SyncProgressCount> progressListener,
+  public boolean execute(TaskProgressListener<SyncProgressCount> progressListener,
       TaskCompletionListener<SyncProgressCount> completionListener) {
     this.progressListener = progressListener;
+    hasErrors = false;
+    syncErrorList = new ArrayList<>();
 
     int pageCount = fetcher.init();
 
-
     List<T> sourceDataList = fetcher.fetchNext();
+    // TODO (@Satya) check if there were errors in fetcher and add to error list.
+
     int currentPage = 0;
     int progressCount = 0;
     int itemsOnPage = 0;
 
     while (sourceDataList != null) {
       List<S> collatedList = collator.collate(sourceDataList);
+      // TODO (@Satya) check if there were errors in collator and add to error list.
       itemsOnPage = collatedList.size();
 
       updateProgress(new SyncProgressCount(++currentPage, pageCount, progressCount, itemsOnPage));
 
       for (S data : collatedList) {
-        R request = requestGenerator.generateSyncRequest(data);
+        R request = null;
+        try {
+          request = requestGenerator.generateSyncRequest(data);
+          // TODO (@Satya) check if there were errors and add them to the error list.
+        } catch (Throwable t) {
+          hasErrors = true;
+          addError(new SyncErrorData(APISyncErrorLocation.API_SYNC_EXECUTE,
+              BaseAPISyncErrorCode.NO_REQUEST_GENERATED, t));
+        }
         if (request != null) {
           try {
             apiExecutor.init(data, request);
           } catch (IOException e) {
-            // TODO (@Satya) if executor could not init - what next?
             logger.info("[SYNC-ERROR] IOError from apiExecutor for request object {} for source {}: " + e.toString(),
                 request.toString(),
                 data.toString());
+            hasErrors = true;
```

Comment:
Again do we need to keep track of the S that caused this error. btw does S have access to the T?

Reply:
we have no defined relationship between S and T. the way we have used it, S is either RequestSource<T> or RequestSource<List<T>>.